### ISSUE TITLE
Custom resources with duplicate names silently fail

### ISSUE THREAD

**Issue by steelcowboy** [chef/chef#9885]
## Description
If two very similar custom resources are created that provide for different platforms, and then they are called in a unified way in a recipe, the resource itself will silently fail: it will say up-to-date on the action but not actually run anything else. 

See [the CPE Chrome default recipe](https://github.com/facebook/IT-CPE/blob/5e0bce5876e88faeb8a4d272566d2627af4423db/itchef/cookbooks/cpe_chrome/recipes/default.rb), and the two resources [cpe_chrome](https://github.com/facebook/IT-CPE/blob/5e0bce5876e88faeb8a4d272566d2627af4423db/itchef/cookbooks/cpe_chrome/resources/cpe_chrome.rb) and [cpe_chrome_win](https://github.com/facebook/IT-CPE/blob/5e0bce5876e88faeb8a4d272566d2627af4423db/itchef/cookbooks/cpe_chrome/resources/cpe_chrome_win.rb) for an example.

## Chef Version
Chef Infra Client: 16.1.0

## Platform Version
Fedora 32

## Replication Case
1. Create a cookbook with two custom resources
2. Give them the same `resource_name`, provide for different platform
3. See that instead of an error, it just says "up to date" and doesn't run the resource

## Client Output
https://gist.github.com/steelcowboy/d411bc080af9b46c11661c96a58e8753

## Stacktrace
N/A

**Comment by btm** [comment]
@steelcowboy as I've commented over on https://github.com/facebook/IT-CPE/issues/243#issuecomment-631064185, I can't reproduce this.

Given this simple reproduction case:
```
btm@btm-thinkpad:/tmp/cookbooks$ find two_resources/
two_resources/
two_resources/resources
two_resources/resources/mirror_windows.rb
two_resources/resources/mirror_linux.rb
two_resources/recipes
two_resources/recipes/default.rb

btm@btm-thinkpad:/tmp/cookbooks$ cat two_resources/recipes/default.rb 
mirror_resource 'testing'

btm@btm-thinkpad:/tmp/cookbooks$ cat two_resources/resources/mirror_linux.rb 
resource_name 'mirror_resource'
provides :mirror_resource, :os => "linux"
default_action :test

action :test do
  file "/tmp/mirror_resource" do
    content "this is linux"
  end
end

btm@btm-thinkpad:/tmp/cookbooks$ cat two_resources/resources/mirror_windows.rb 
resource_name 'mirror_resource'
provides :mirror_resource, :os => "windows"
default_action :test

action :test do
  file "/tmp/mirror_resource" do
    content "this is windows"
  end
end
```

As far as I can tell this is working correctly:
```
btm@btm-thinkpad:/tmp/cookbooks$ chef-client -z -o two_resources
[2020-05-19T16:39:31-04:00] WARN: No config file found or specified on command line. Using command line options instead.
Starting Chef Infra Client, version 16.0.287
[2020-05-19T16:39:34-04:00] ERROR: shard_seed: Failed to get dmi property serial_number: is dmidecode installed?
[2020-05-19T16:39:35-04:00] WARN: Run List override has been provided.
[2020-05-19T16:39:35-04:00] WARN: Original Run List: []
[2020-05-19T16:39:35-04:00] WARN: Overridden Run List: [recipe[two_resources]]
resolving cookbooks for run list: ["two_resources"]
Synchronizing Cookbooks:
  - two_resources (0.0.0)
Installing Cookbook Gems:
Compiling Cookbooks...
Converging 1 resources
Recipe: two_resources::default
  * mirror_resource[testing] action test
    * file[/tmp/mirror_resource] action create
      - create new file /tmp/mirror_resource
      - update content in file /tmp/mirror_resource from none to 63e3ef
      --- /tmp/mirror_resource	2020-05-19 16:39:35.183393119 -0400
      +++ /tmp/.chef-mirror_resource20200519-373-opetv1	2020-05-19 16:39:35.179393101 -0400
      @@ -1 +1,2 @@
      +this is linux
  
[2020-05-19T16:39:35-04:00] WARN: Skipping final node save because override_runlist was given

Running handlers:
Running handlers complete
Chef Infra Client finished, 2/2 resources updated in 03 seconds
```

Does this work for you?

**Comment by btm** [comment]
@steelcowboy okay, it is not silently failing. It is choosing the windows version of the cpe_chrome resource, and it is returning at the beginning of the action based on one of these conditionals:

https://github.com/facebook/IT-CPE/blob/5e0bce5876e88faeb8a4d272566d2627af4423db/itchef/cookbooks/cpe_chrome/resources/cpe_chrome_win.rb#L27-L28

If I comment out the provides line in my Linux resource from my reproduction case above, then run chef-client again, I get the windows resource:
```
btm@btm-thinkpad:/tmp/cookbooks$ cat two_resources/resources/mirror_linux.rb 
resource_name 'mirror_resource'
#provides :mirror_resource, :os => "linux"
default_action :test

action :test do
  file "/tmp/mirror_resource" do
    content "this is linux"
  end

btm@btm-thinkpad:/tmp/cookbooks$ chef-client -z -o two_resources
[2020-05-19T16:45:29-04:00] WARN: No config file found or specified on command line. Using command line options instead.
Starting Chef Infra Client, version 16.0.287
[2020-05-19T16:45:32-04:00] ERROR: shard_seed: Failed to get dmi property serial_number: is dmidecode installed?
[2020-05-19T16:45:32-04:00] WARN: Run List override has been provided.
[2020-05-19T16:45:32-04:00] WARN: Original Run List: []
[2020-05-19T16:45:32-04:00] WARN: Overridden Run List: [recipe[two_resources]]
resolving cookbooks for run list: ["two_resources"]
Synchronizing Cookbooks:
  - two_resources (0.0.0)
Installing Cookbook Gems:
Compiling Cookbooks...
Converging 1 resources
Recipe: two_resources::default
  * mirror_resource[testing] action test
    * file[/tmp/mirror_resource] action create
      - update content in file /tmp/mirror_resource from 63e3ef to bed3ef
      --- /tmp/mirror_resource	2020-05-19 16:39:35.179393101 -0400
      +++ /tmp/.chef-mirror_resource20200519-1579-1mco8zm	2020-05-19 16:45:32.572971478 -0400
      @@ -1,2 +1,2 @@
      -this is linux
      +this is windows
  
[2020-05-19T16:45:32-04:00] WARN: Skipping final node save because override_runlist was given

Running handlers:
Running handlers complete
Chef Infra Client finished, 2/2 resources updated in 02 seconds
```

**Comment by lamont-granquist** [comment]
Yeah we went round and around on this with the removal of the canonical DSL in #9441 

The Pre-Chef-16 behavior is that the `resource_name :cpe_chrome` call has always silently done a `provides :cpe_chrome, canonical: true` call in the background.  When the second call to `provides :cpe_chrome, os: windows` is called then it would magically notice the presence of the canonical DSL and remove and replace it.

With the canonical DSL gone, there's no way for the second provides line to determine the magical wiring up of the `provides` line caused by setting the resource_name.

My initial approach was to rip out the magical wiring of the `resource_name` to the canonical DSL, such that setting the resource_name...  would only set the resource_name.  Very simple and obvious behavior.

The problem with this is that there are resources out there which only set the `resource_name` and do not set any `provides` line at all and doing that would break them.

So we're here with the behavior in Chef 16 where setting the resource_name magically wires up the provides line and generates this issue.

I'd prefer to go the other direction and drop the `resource_name` wiring up `provides` at all, and just have it be the `resource_name`.   That would mean that any resource not declaring a `provides` line would be broken.  But that, at least, is fully obvious what is going on.

(There is also some minimal magic in Chef-16 where if the resource_name hasn't been set it gets set by the first provides line, with later `resource_name` calls overriding it -- that feels like the right direction for magic to flow though since `resource_name` is semantically a lot "smaller" concept while `provides` is a much "larger" concept -- but that does mean that you can call one provides line and omit the resource_name entirely).

**Comment by lamont-granquist** [comment]
closed by #9890
